### PR TITLE
Update search markup

### PR DESCRIPTION
**Pull request by owenatgov**
## What
Update the markup, classnames and styling on the search view and associated components (filters and sort). There should be no discernible difference between this PR and production.

## Why
DGU Find is currently using govuk elements, an outdated implementation of the govuk design system. This PR is one of several that attempts to slowly remove the elements implementation and replace it with up to date implementations from govuk frontend and govuk publishing components.

**Card:** https://trello.com/c/EbYBZqRv/232-update-markup-on-find-search-page

### CODE REVIEW

File: app/assets/stylesheets/govuk-overrides.scss [hunk 1]
```diff
@@ -86,3 +86,7 @@ button.button.secondary {
     background-image: image-url("separator-2x.png");
   }
 }
+
+.gem-c-search--on-white .gem-c-search__submit {
```
Comment (andysellick) [inline app/assets/stylesheets/govuk-overrides.scss:90]: What's the reason for this override?

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -23,6 +23,7 @@ $govuk-use-legacy-palette: false;
 @import 'govuk_publishing_components/components/_layout-header';
 @import 'govuk_publishing_components/components/_layout-footer';
 @import 'govuk_publishing_components/components/_title';
+@import 'govuk_publishing_components/components/_search';
```
Comment (andysellick) [inline app/assets/stylesheets/application.scss:26]: This is the stylesheet for the search component, but I can't see the component being called in the code. Is it in a separate commit or already present somewhere?

File: app/views/search/_filters.html.erb [hunk 3]
```diff
@@ -1,32 +1,32 @@
 <h2 class="heading-medium"> <%= t('.filter_by') %></h2>
 
-<div class="form-group">
-  <%= label_tag 'publisher', t('.filter_by_publisher'), class: 'form-label' %>
-  <%= select_tag('filters[publisher]', options_for_select(dataset_publishers_for_select, selected_publisher), id: 'publisher', class: 'form-control dgu-filters__filter', include_blank: true) %>
+<div class="govuk-form-group">
+  <%= label_tag 'publisher', t('.filter_by_publisher'), class: 'govuk-label' %>
+  <%= select_tag('filters[publisher]', options_for_select(dataset_publishers_for_select, selected_publisher), id: 'publisher', class: 'govuk-select dgu-filters__filter', include_blank: true) %>
```
Comment (andysellick) [inline app/views/search/_filters.html.erb:5]: This looks like the markup for the select component, is that right? Why is it written out rather than calling the component? e.g. 

```
<%= render "govuk_publishing_components/components/select", {
  id: "dropdown1",
  label: "My Dropdown",
  options: [
    {
      text: "Option one",
      value: "option1"
    },
  ]
} %>

File: app/views/search/_sort.html.erb [hunk 4]
```diff
@@ -1,11 +1,11 @@
 <div class="dgu-sort">
-  <div class="dgu-sort__apply visuallyhidden">
+  <div class="dgu-sort__apply govuk-visually-hidden">
     <input type="submit" value=<%= t('.accessibility.submit_button') %>/>
   </div>
 
   <div class="dgu-sort__sort-type">
     <label>
-      <span class="visuallyhidden"> <%= t('.accessibility.sort_type_label') %></span>
+      <span class="govuk-visually-hidden"> <%= t('.accessibility.sort_type_label') %></span>
```
Comment (andysellick) [inline app/views/search/_sort.html.erb:8]: This PR deployment isn't working, and I don't have context of the UI, why is this label text being visually hidden?

Comment (andysellick) [inline app/views/search/_sort.html.erb:8]: If you want to visually hide the label, could you put the `govuk-visually-hidden` class on the label itself?

File: app/assets/stylesheets/govuk-overrides.scss (hunk 1 above)
Comment (owenatgov) [reply app/assets/stylesheets/govuk-overrides.scss:90]: The button generated by the GUPC instance of the search bar is blue whilst the search bar button across data.gov.uk currently is black. I'm not sure if it makes more sense to just leave it blue, if there's been research done on this etc. The team's in an awkward position where we can't really get design support so I don't want to make any non-trivial changes. If you think I could get away with a blue button I'm happy to chat to the team and scoop this out.

File: app/views/search/_filters.html.erb (hunk 3 above)
Comment (owenatgov) [reply app/views/search/_filters.html.erb:5]: This is an accessible autocomplete instance. I wasn't sure if it was possible to include the select component here and for the AA to still work. I will try this out and get back to you.

File: app/views/search/_sort.html.erb (hunk 4 above)
Comment (owenatgov) [reply app/views/search/_sort.html.erb:8]: It's an accessible label that's sitting next to a search filter. You can see it on this page https://data.gov.uk/search?q= next to the filter dropdown with the selection "best match". I will however make the label change you noted, good spot.

File: app/assets/stylesheets/application.scss (hunk 2 above)
Comment (andysellick) [reply app/assets/stylesheets/application.scss:26]: Ah, sorry, never mind then.

File: app/views/search/_filters.html.erb (hunk 3 above)
Comment (andysellick) [reply app/views/search/_filters.html.erb:5]: It probably isn't, so this is probably fine as it is. It wasn't obvious that an AA is applied here, is that done through a specific class? I think it might be worth making this more obvious, perhaps by using a specific `js-` prefixed class for example.

File: app/assets/stylesheets/govuk-overrides.scss (hunk 1 above)
Comment (andysellick) [reply app/assets/stylesheets/govuk-overrides.scss:90]: One of the principles of our components is that they're isolated, which means we can have confidence that if it looks correct in the component guide, it looks correct everywhere. If we style components in applications like this, it can lead to problems - for example, if the component was updated to use `govuk-text-colour` as the text colour, in your application the text on the button would suddenly become unreadable. That's not likely to happen, but the rule is there to defend against situations that are more plausible.

The other advantage of not styling components in applications is that we get a consistent look across the site. Is there a reason why DGU has subtle differences like this?

File: app/views/search/_sort.html.erb (hunk 4 above)
Comment (andysellick) [reply app/views/search/_sort.html.erb:8]: I'm still interested to know why the label is hidden. If you look at https://www.gov.uk/search/all?keywords=minister&order=relevance there's a similar 'sort by' option (beneath the result count) and the label is shown.

File: app/assets/stylesheets/govuk-overrides.scss (hunk 1 above)
Comment (owenatgov) [reply app/assets/stylesheets/govuk-overrides.scss:90]: I can't dispute this. The question has been asked, will report back.

File: app/views/search/_sort.html.erb (hunk 4 above)
Comment (owenatgov) [reply app/views/search/_sort.html.erb:8]: I'm guessing it was just a design choice 😬 I'll re-work it.

File: app/views/search/_filters.html.erb (hunk 3 above)
Comment (owenatgov) [reply app/views/search/_filters.html.erb]: I haven't been able to get this working with the select govpubcomp but I've added a js class to label it so it's clear that it's being used for AA.

File: app/assets/javascripts/govuk.js [hunk 5]
```diff
@@ -15,28 +15,24 @@ $(document).ready(function () {
   var showHide = new ShowHide()
   showHide.init()
 
-  if (document.getElementById('publisher')) {
-    accessibleAutocomplete.enhanceSelectElement({
-      selectElement: document.querySelector('#publisher'),
-      showAllValues: true,
-      preserveNullOptions: true
-    })
+  var autocompleteSelects = document.getElementsByClassName('js-accessible-autocomplete');
```
Comment (injms) [inline app/assets/javascripts/govuk.js:18]: From the look of the rest of this file, it looks as if it's been linted with [StandardJS](http://standardjs.com), which is the chosen linter from the [GDS Way rules for JavaScript](https://gds-way.cloudapps.digital/manuals/programming-languages/js.html#linting).

I think the new bits of JavaScript need to be linted as well to ensure that the file is consistent.

```suggestion
  var autocompleteSelects = document.getElementsByClassName('js-accessible-autocomplete')
```

File: app/views/search/_filters.html.erb [hunk 6]
```diff
@@ -1,32 +1,32 @@
 <h2 class="heading-medium"> <%= t('.filter_by') %></h2>
 
-<div class="form-group">
-  <%= label_tag 'publisher', t('.filter_by_publisher'), class: 'form-label' %>
-  <%= select_tag('filters[publisher]', options_for_select(dataset_publishers_for_select, selected_publisher), id: 'publisher', class: 'form-control dgu-filters__filter', include_blank: true) %>
+<div class="govuk-form-group">
+  <%= label_tag 'publisher', t('.filter_by_publisher'), class: 'govuk-label' %>
```
Comment (injms) [inline app/views/search/_filters.html.erb:4]: If DGU uses it, the [GDS Rubocop rules](https://gds-way.cloudapps.digital/manuals/programming-languages/ruby.html#code-formatting) ask for double quotes. 

It can be run to automatically fix some things, so these might be better in another pull request.
```suggestion
  <%= label_tag "publisher", t(".filter_by_publisher"), class: "govuk-label" %>
```

File: app/views/search/_sort.html.erb [hunk 7]
```diff
@@ -1,15 +1,22 @@
...
+      },
+      {
+        text: t('.recent'),
+        value: 'recent',
+        selected: @sort === 'recent'
+      }
+    ]
+  } %>
+  <div class="govuk-visually-hidden">
     <input type="submit" value=<%= t('.accessibility.submit_button') %>/>
```
Comment (injms) [inline app/views/search/_sort.html.erb:20]: This button could be a button component:

```suggestion
    <%= render "govuk_publishing_components/components/button", {
        text: t('.accessibility.submit_button'),
    } %>
```

File: app/views/search/search.html.erb [hunk 8]
```diff
@@ -3,81 +3,74 @@
...
-    </div>
-  </div>
-
-  <div class="grid-row dgu-filters">
-    <div class="column-one-third">
-      <%= render 'filters' %>
-    </div>
+  <main role="main" id="main-content" class="govuk-main-wrapper">
+    <div class="govuk-width-container">
+      <h1 class="govuk-heading-l"> <%= t('.search_results_heading') %></h1>
```
Comment (injms) [inline app/views/search/search.html.erb:8]: This could be a [heading component](https://components.publishing.service.gov.uk/component-guide/heading):
```suggestion
      <%= render "govuk_publishing_components/components/heading", {
        text: t('.search_results_heading'),
        heading_level: 1,
        font_size: "l",
      } %>
```

File: app/views/search/search.html.erb [hunk 9]
```diff
@@ -3,81 +3,74 @@
...
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds dgu-search-box">
```
Comment (injms) [inline app/views/search/search.html.erb:10]: Could `dgu-search-box` be removed here as the search component is being used?
```suggestion
        <div class="govuk-grid-column-two-thirds">
```

File: app/views/search/search.html.erb [hunk 10]
```diff
@@ -3,81 +3,74 @@
...
-              <li><%= t('.zero_result_tips.different_words') %></li>
-              <li><%= t('.zero_result_tips.remove_filters') %></li>
-            </ul>
-            <h2 class="heading-medium"> <%= t('.zero_result_tips.older_content') %></h2>
-            <p><%= t('.zero_result_tips.we_can_help') %></p>
-            <p><%= t('.zero_result_tips.expand_your_search') %></p>
+        <div class="govuk-grid-column-two-thirds dgu-results">
+          <%= render 'sort' if @num_results > 1 %>
+          <span class="dgu-results__summary">
+            <span class="bold-small"><%= number_with_delimiter(@num_results) %></span>
```
Comment (injms) [inline app/views/search/search.html.erb:25]: We could swap this bespoke CSS out for some styles from the Design System:

```suggestion
            <span class="govuk-body-s govuk-!-font-weight-bold"><%= number_with_delimiter(@num_results) %></span>
```

File: app/views/search/search.html.erb [hunk 11]
```diff
@@ -3,81 +3,74 @@
...
-                <% if dataset.public_updated_at.present? %>
-                  <dd class="last_updated"><%= format_timestamp(dataset.public_updated_at) %></dd>
-                <% else %>
-                  <dd class="last_updated dgu-secondary-text"> <%= t('.meta_data_box.not_applicable') %></dd>
-                <% end %>
-              </dl>
-              <p><%= truncate(strip_tags(dataset.summary), length: 200, separator: ' ') %></p>
+          <% if @num_results.zero? %>
+            <div class="dgu-results__zero">
+              <h2 class="govuk-heading-m"> <%= t('.zero_result_tips.try') %>:</h2>
```
Comment (injms) [inline app/views/search/search.html.erb:31]: Another potential heading component:
```suggestion
          <%= render "govuk_publishing_components/components/heading", {
                text: t('.zero_result_tips.try'),
                heading_level: 2,
                font_size: "m",
              } %>
              <h2 class="govuk-heading-m"> <%= t('.zero_result_tips.try') %>:</h2>
```

File: app/views/search/search.html.erb [hunk 12]
```diff
@@ -3,81 +3,74 @@
...
+              <ul class="govuk-list govuk-list--bullet">
+                <li><%= t('.zero_result_tips.different_words') %></li>
+                <li><%= t('.zero_result_tips.remove_filters') %></li>
+              </ul>
+              <h2 class="govuk-heading-m"> <%= t('.zero_result_tips.older_content') %></h2>
```
Comment (injms) [inline app/views/search/search.html.erb:36]:
```suggestion
              <%= render "govuk_publishing_components/components/heading", {
                text: t(".zero_result_tips.older_content"),
                heading_level: 2,
                font_size: "m",
              } %>
```

**Comment by owenatgov** [comment]
Rubocop is failing on a few things but I'm going to propose that we bypass this as most of the problems are going to be solved by https://github.com/alphagov/datagovuk_find/pull/790

**Comment by kentsanggds** [comment]
I think the upgrade to Rails 6 and linting fixes have been merged into master so you probably just need to rebase to get tavis to pass
